backends/bmv2/jsonconverter.cpp

```diff
@@ -1216,6 +1216,134 @@ Util::IJson* JsonConverter::convertIf(const CFG::IfNode* node, cstring) {
+                    auto count_ones = [](unsigned long n) { return n ? __builtin_popcountl(n) : 0; };
+                    unsigned long mask = k->to<IR::Mask>()->right->to<IR::Constant>()->value.get_ui();
+                    auto len = trailing_zeros(mask);
+                    if(len + count_ones(mask) != keyWidth) // any remaining 0s in the prefix?
+                        ::error("%1% invalid mask for LPM key", k);
+                    else
+                        key->emplace("prefix_length", keyWidth - len);
+                } else if (k->is<IR::Constant>()) {
+                    key->emplace("key", stringRepr(k->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                    key->emplace("prefix_length", keyWidth);
+                } else if (k->is<IR::DefaultExpression>()) {
+                    key->emplace("key", stringRepr(0, ROUNDUP(keyWidth, 8)));
+                    key->emplace("prefix_length", 0);
+                } else {
+                    ::error("%1% invalid LPM key expression", k);
+                }
+            } else if (matchType == "range") {
+                if (k->is<IR::Range>()) {
+                    key->emplace("start", stringRepr(k->to<IR::Range>()->left->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                    key->emplace("end", stringRepr(k->to<IR::Range>()->right->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                } else if (k->is<IR::DefaultExpression>()) {
+                    key->emplace("start", stringRepr(0, ROUNDUP(keyWidth, 8)));
+                    key->emplace("end", stringRepr((1<<keyWidth)-1, ROUNDUP(keyWidth, 8))); // 2^N -1
+                } else {
+                    ::error("%1% invalid range key expression", k);
+                }
+            } else {
+                ::error("unkown key type %1% for key %2%", matchType, k);
+            }
+            matchKeys->append(key);
+            keyIndex++;
+        }
+
+        auto action = new Util::JsonObject();
+        auto actionRef = e->getAction();
+        if (!actionRef->expression->is<IR::MethodCallExpression>())
+            ::error("%1%: invalid action in entries list", actionRef);
+        auto actionCall = actionRef->expression->to<IR::MethodCallExpression>();
+        auto decl = refMap->getDeclaration(actionRef->getPath(), true);
+        auto actionDecl = decl->to<IR::P4Action>();
+        unsigned id = get(structure.ids, actionDecl);
+        action->emplace("action_id", id);
+        auto actionData = mkArrayField(action, "action_data");
+        for (auto arg: *actionCall->arguments) {
+            actionData->append(stringRepr(arg->to<IR::Constant>()->value, 0));
+        }
+        entry->emplace("action_entry", action);
+
+        if (e->getAnnotations() != IR::Annotations::empty) {
+            for (auto p: e->getAnnotations()->annotations) {
+                if (p->name == "priority") {
```

Comment:
Maybe we should make "priority" a static constant; if we put it in the spec, it should be in corelib.